accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,493 @@
+
+The behavior is *only* for the `this` operator. Doing something like `var self = this; self.isEven;` will not give preference over members of `List` (if `List` had an `isEven`, then `self.isEven` would invoke that, where `this.isEven` will invoke the extension method). It is as if `this` has a type representing the extension, even if no such type actually exists.
+
+This behavior makes the extension act similarly to a class, which will hopefully make it easier for users to understand the resolution.
+
+Like for a class or mixin member declaration, the names of the extension members, both static and instance, are in the *lexical* scope of the extension member. That is why `MySmart` above can invoke the static `smartHelper` without prefixing it by the extension name. In the same way, *instance* members are in the lexical scope. 
+
+Example:
+
+```dart
+extension MyUnaryNumber on List<Object> {
+  bool get isEven => this.length.isEven;
+  bool get isOdd => !isEven;  // not `!this.isEven`
+}
+```
+
+Here, the `isEven` name exists in the surrounding static scope, so just as for `class` members, it is considered equivalent to `this.isEven`. Because of how we treat `this`, it means that `isEven` and `this.isEven` means the same thing when the extension declares an `isEven` member, just as for a class or mixin declaration.
+
+### Member Conflict Resolution
+
+An extension can declare a member with the same (base-)name as the type it is declared on. This does not cause a compile-time conflict, even if the member does not have a compatible signature.
+
+Example:
+
+```dart
+extension MyList<T> on List<T> {
+  void add(T value, {int count = 1}) { ... }
+  void add2(T value1, T value2) { ... }
+}
+```
+
+You cannot *access* this member in a normal invocation, so it could be argued that you shouldn't be allowed to add it. We allow it because we do not want to make it a compile-time error for a type to add a method just because an extension is already adding a method with the same name. it will likely be a problem if any code *uses* the method, but only that code needs to change (perhaps using an override).
+
+That also means that an extension can shadow members of the `on` type.  To implement `add2` above, we don't want to call `this.add`, instead we want to use the `add` of `List` directly. To do that, we allow `super` invocations to target the underlying `on` type directly, with *no* extensions applying.
+
+```dart
+void add2(T value1, T value2) {
+  super..add(value1)..add(value2);
```

Comment:
I'm worried this encourages a mental that really isn't correct. Extensions aren't on a magic subclass of the type. If they were, then a non-self call would hit the extension first and not the instance method.

Are you sure this `super()` stuff is needed at all?

Reply:
The `super` stuff is no longer present in the proposal (except for a tiny bit in the sections about variants), so I believe we can resolve this thread.